server/routes/appointments/create/handlers/uploadPrisonerList.ts

```diff
@@ -0,0 +1,101 @@
+
+export class PrisonerList {
+  @Expose()
+  @IsNotEmpty({ message: "Select a CSV file of prisoners' numbers" })
+  @IsNotEmptyFile({ message: 'The selected file is empty' })
+  @IsValidCsvFile({ message: 'The selected file must be a CSV' })
+  file: Express.Multer.File
+}
+
+export default class UploadPrisonerListRoutes {
+  constructor(private readonly prisonService: PrisonService) {}
+
+  GET = async (req: Request, res: Response): Promise<void> => {
+    res.render(`pages/appointments/create/upload-prisoner-list`)
+  }
+
+  POST = async (req: Request, res: Response): Promise<void> => {
+    const prisonerListCsvFile = req.file
+    const { user } = res.locals
+
+    const prisonerNumbers: string[] = []
+    const errors: string[] = []
+
+    fs.readFile(prisonerListCsvFile.path, (err, result) => {
+      if (err) {
+        req.flash(
+          'validationErrors',
+          JSON.stringify([{ field: 'file', message: 'The selected file could not be uploaded – try again' }]),
+        )
+        return res.redirect('back')
+      }
+
+      fs.unlinkSync(prisonerListCsvFile.path)
+
+      const parser = parse(result, { trim: true, skipEmptyLines: true })
+        .on('readable', () => {
+          let row
+          // eslint-disable-next-line no-cond-assign
+          while ((row = parser.read())) {
+            const prisonerNumber = row[0]
+            if (prisonerNumber !== 'Prison number' && !prisonerNumbers.includes(prisonerNumber)) {
+              prisonerNumbers.push(prisonerNumber)
+            }
+          }
+        })
+        .on('error', error => {
+          errors.push(error.message)
+        })
+        .on('end', async () => {
+          if (errors.length > 0) {
+            req.flash(
+              'validationErrors',
+              JSON.stringify([{ field: 'file', message: 'The selected file must use the template' }]),
+            )
+            return res.redirect('back')
+          }
+
+          const prisoners = (await this.prisonService.searchInmatesByPrisonerNumbers(prisonerNumbers, user)).map(
+            prisoner => ({
+              number: prisoner.prisonerNumber,
+              name: `${prisoner.firstName} ${prisoner.lastName}`,
+              cellLocation: prisoner.cellLocation,
+            }),
+          )
+
+          const prisonerNumbersFound = prisoners.map(prisoner => prisoner.number)
+
+          const prisonerNumbersNotFound = prisonerNumbers.filter(
+            prisonerNumber => !prisonerNumbersFound.includes(prisonerNumber),
+          )
+
+          if (prisonerNumbersNotFound.length > 0) {
+            const message =
+              prisonerNumbersNotFound.length === 1
+                ? `Prisoner with number ${prisonerNumbersNotFound[0]} was not found`
+                : `Prisoners with numbers ${prisonerNumbersNotFound.join(', ')} were not found`
+            req.flash('validationErrors', JSON.stringify([{ field: 'file', message }]))
+            return res.redirect('back')
```

Comment:
If there's more than one prisoner number that isn't found, it will present one validation message containing all the invalid numbers. The other option is adding the ones that aren't found then having a warning message on the next page that "we couldn't find the following prisoner numbers: xxxxxx" kind of thing. I think that's a nicer pattern and have suggested it

Reply:
Long :). It does do a distinct on the supplied numbers so they'd have to upload 1000 different numbers all of which weren't recognised. Enough of an edge case to ignore or would you prefer something like if there's greater than 10 invalid numbers we say "XXX prisoner numbers were not found"?